torch/csrc/jit/codegen/cuda/lower_trivial_reductions.cpp

```diff
@@ -23,29 +23,46 @@ bool traverseToRFactorTensor(TensorView* tv, IterDomain* root_id) {
   TORCH_INTERNAL_ASSERT(
       root_id->definition() == nullptr, "Not root IterDomain: ", root_id);
 
-  if (tv->definition() == nullptr) {
+  auto def = tv->definition();
+
+  if (def == nullptr) {
     // This is an input tensor, so no rfactor tensor to traverse.
     return false;
   }
 
-  const auto& inputs = tv->definition()->inputs();
-
   // Check the reduction expression that produces tv
-  if (inputs.size() != 1 || !inputs[0]->isA<TensorView>() ||
-      (tv->definition()->getExprType() != ExprType::ReductionOp &&
-       tv->definition()->getExprType() != ExprType::WelfordOp)) {
-    // No rfactor producer found
+  if (!ir_utils::isReductionOp(def)) {
     return false;
   }
 
-  auto producer = inputs[0]->as<TensorView>();
+  // Find the corresponding input TV. Note that the reduction expr may
+  // have multiple inputs.
+  auto producer = def->inputs().at(std::distance(
+      def->outputs().begin(),
+      std::find(def->outputs().begin(), def->outputs().end(), tv)));
+
+  auto producer_tv = dynamic_cast<TensorView*>(producer);
+
+  // WelfordOp may have an Int input. Traverse to the avg input
+  if (def->isA<WelfordOp>() && producer_tv == nullptr) {
```

Comment:
Why do we need to grab the "right" producer? Can't we just take the first TV input? They should have to be aligned to be siblings.

Reply:
That should be fine with `WelfordOp`, but in `GroupedReductionOp`, in theory, the input tensors just need to have the same shape. It should be fine for some of them to have rfactor domains, although the current validation may not be flexible enough to accept such a case. So, picking the right input could be important.

Reply:
How would some reductions have rfactor and others not with grouped reduction? I assume you'd have to have some interesting view op in the dag?

Reply:
Will revisit this again.

Reply:
`groupReductions` can group an arbitrary set of `ReductionOp` exprs as long as they have the same input shape. I don't know if it could ever happen in practice, but it is in theory possible to group a reduction of a post-view tensor and another reduction of a tensor that has the same shape as the post-view tensor.